src/auth/awsclient.ts

```diff
@@ -335,4 +341,36 @@ export class AwsClient extends BaseExternalAccountClient {
     });
     return response.data;
   }
+
+  private shouldUseMetadataServer(): boolean {
+    // The metadata server must be used when either the AWS region or AWS security
+    // credentials cannot be retrieved through their defined environment variables.
+    return (
+      !this.regionFromEnvironment || !this.securityCredentialsFromEnvironment
+    );
+  }
+
+  private get regionFromEnvironment(): string | undefined {
+    // The AWS region can be provided through AWS_REGION or AWS_DEFAULT_REGION.
+    // Only one is required.
+    return process.env['AWS_REGION'] || process.env['AWS_DEFAULT_REGION'];
+  }
+
+  private get securityCredentialsFromEnvironment():
+    | AwsSecurityCredentialsFromEnvironment
+    | undefined {
+    let awsSecurityCredentials;
+    // Check if both AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are available.
+    if (
+      process.env['AWS_ACCESS_KEY_ID'] &&
+      process.env['AWS_SECRET_ACCESS_KEY']
+    ) {
+      awsSecurityCredentials = {
+        accessKeyId: process.env['AWS_ACCESS_KEY_ID'],
+        secretAccessKey: process.env['AWS_SECRET_ACCESS_KEY'],
+        token: process.env['AWS_SESSION_TOKEN'],
+      };
+    }
+    return awsSecurityCredentials;
+  }
```

Comment:
This could lead to a use before reference error if an environment variable isn't available. Here's an alternative:

```suggestion
  private get securityCredentialsFromEnv(): AwsSecurityCredentials | null {
    // Both AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are required.
    if (
      process.env['AWS_ACCESS_KEY_ID'] &&
      process.env['AWS_SECRET_ACCESS_KEY']
    ) {
      return {
        accessKeyId: process.env['AWS_ACCESS_KEY_ID'],
        secretAccessKey: process.env['AWS_SECRET_ACCESS_KEY'],
        token: process.env['AWS_SESSION_TOKEN'],
      };
    }

    return null;
  }
```